### DISCUSSION TITLE
Unable to remove Cam controller from Renderer (?)

### DISCUSSION THREAD

**Discussion by SteKme** [pygfx/pygfx#431] [Ideas]
Hi,

so far I don't see any way to remove a controller from the render - in Controller's add_default_event_handlers(), lambda is registered, however to remove it from the renderer, one'd need reference to it (base.py/remove_event_handler()?)

It is normal to change number of viewports over time, but so far, the best I came up with is to move a viewport associated with the camera+controller to some rect coordinates where viewport's is_inside() never succeeds to get rid of it. Could there be a better way, please? E.g. Blender - have 2x2 layout, then switch to a single viewport, then return back to 2x2 etc..

Also, it is a shame these don't register with the renderer aside of the callback, so I don't see any way to enumerate them from the renderer and must save references myself separately (applies to controllers, viewports..).

Thank You for all valuable efforts and best regards,
Stepan

**Comment by Korijn** [comment]
To be clear, the controllers are a sort of "bonus" feature, to be used in examples, demos and similar minimal applications. At its core, pygfx is just the library you use for rendering. The GUI layer and interaction code is not included, you are expected to implement that as part of your own application code.

Nevertheless, you can propose an improved API and open a Pull Request and we can review your changes. As a starting point, you can add a function to controller to detach all event handlers.
